File: lua/binfmt.lua
local msg =
"this is binary image of https://github.com/bsahd/image-compress format.\nversion:230606ee9a6d0b45b71167f8faa01ed169cd96bb\n\n\n\n\n\n\n\n\n"
local endmsg = "\n\n\nthis is binary format. read head using head command for more information.\n"
local struct
if string.pack then
    struct = { pack = string.pack, unpack = string.unpack }
else
    struct = require("struct")
end

local _unpack
if table.unpack then
    _unpack = table.unpack
else
    _unpack = unpack
end

local M = {}
function M.buf_to_img(data)
    local img = {}
    local read_head = 1

    local function read_buf(fmt)
        local results_from_unpack = { struct.unpack(fmt, data, read_head) }
        local next_pos = nil
        local actual_values = {}
        if #results_from_unpack == 0 then
            error("Failed to unpack data for format: " .. fmt .. " at position: " .. read_head .. ". No values returned.")
        end
        next_pos = results_from_unpack[#results_from_unpack]
        for i = 1, #results_from_unpack - 1 do
            actual_values[i] = results_from_unpack[i]
        end
        read_head = next_pos
        return _unpack(actual_values)
    end

    local header = read_buf("c" .. #msg)
    assert(header == msg, "Invalid header")

    img.width = read_buf(">i2")
    img.height = read_buf(">i2")
    local blockcount = read_buf(">i4")
    img.blocks = {}

    for i = 1, blockcount do
        local elem = {}
        elem.blockmaxy, elem.blockminy, elem.blockmaxu, elem.blockminu, elem.blockmaxv, elem.blockminv = read_buf(
            ">BBBBBB")
        local interpolaten = read_buf(">B")
        elem.interpolatey = interpolaten >= 4
        elem.interpolateu = (interpolaten % 4) >= 2
        elem.interpolatev = (interpolaten % 2) == 1
        elem.corners = {}
        elem.nblock4bn = {}
        table.insert(img.blocks, elem)
    end

    for i = 1, blockcount do
        if not img.blocks[i] then error("Block " .. i .. " not found in img.blocks") end
        for _ = 1, 4 do
            table.insert(img.blocks[i].corners, read_buf(">B"))
        end
    end

    for i = 1, blockcount do
        if not img.blocks[i] then error("Block " .. i .. " not found in img.blocks for nblock4bn") end
        for _ = 1, 8 do
            local row = {}
            for _ = 1, 8 do
                table.insert(row, read_buf(">B"))
            end
            table.insert(img.blocks[i].nblock4bn, row)
        end
    end

    local footer = read_buf("c" .. #endmsg)
    assert(footer == endmsg, "Invalid footer")

    return img
end

function M.img_to_buf(img)
    local parts = {}
    table.insert(parts, msg)
    table.insert(parts, struct.pack(">i2", img.width))
    table.insert(parts, struct.pack(">i2", img.height))
    table.insert(parts, struct.pack(">i4", #img.blocks))

    for _, block in ipairs(img.blocks) do
        table.insert(parts,
            struct.pack(">BBBBBB", block.blockmaxy, block.blockminy, block.blockmaxu, block.blockminu, block.blockmaxv,
                block.blockminv))
        local interpolaten = (block.interpolatey and 4 or 0) + (block.interpolateu and 2 or 0) +
            (block.interpolatev and 1 or 0)
        table.insert(parts, struct.pack(">B", interpolaten))
    end

    for _, block in ipairs(img.blocks) do
        for _, corner in ipairs(block.corners) do
            table.insert(parts, struct.pack(">B", corner))
        end
    end

    for _, block in ipairs(img.blocks) do
        for _, row in ipairs(block.nblock4bn) do
            for _, p in ipairs(row) do
                table.insert(parts, struct.pack(">B", p))
            end
        end
    end

    table.insert(parts, endmsg)
    return table.concat(parts)
end

return M

File: c/binfmt.c
#include "binfmt.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
#include <winsock2.h>
#else
#include <arpa/inet.h>
#endif

static const char *binfmt_msg =
    "this is binary image of https://github.com/bsahd/image-compress "
    "format.\nversion:"
    "230606ee9a6d0b45b71167f8faa01ed169cd96bb\n\n\n\n\n\n\n\n\n";
static const char *binfmt_endmsg = "\n\n\nthis is binary format. read head "
                                   "using head command for more information.\n";

void img_to_buf(const ImgData *imgdata, char **buffer, size_t *size) {

  size_t total_size = strlen(binfmt_msg) + strlen(binfmt_endmsg) +
                      sizeof(int16_t) * 2 + sizeof(int32_t) +
                      imgdata->block_count * (sizeof(uint8_t) * 7);
  total_size += imgdata->block_count * (sizeof(uint8_t) * 4);
  total_size += imgdata->block_count * (sizeof(uint8_t) * 64);

  char *buf = (char *)malloc(total_size);
  if (!buf) {
    fprintf(stderr, "Memory allocation failed for binfmt buffer.\n");
    *buffer = NULL;
    *size = 0;
    return;
  }
  char *ptr = buf;

  memcpy(ptr, binfmt_msg, strlen(binfmt_msg));
  ptr += strlen(binfmt_msg);

  int16_t width_be = htons(imgdata->width);
  int16_t height_be = htons(imgdata->height);
  memcpy(ptr, &width_be, sizeof(int16_t));
  ptr += sizeof(int16_t);
  memcpy(ptr, &height_be, sizeof(int16_t));
  ptr += sizeof(int16_t);

  int32_t block_count_be = htonl(imgdata->block_count);
  memcpy(ptr, &block_count_be, sizeof(int32_t));
  ptr += sizeof(int32_t);

  for (int i = 0; i < imgdata->block_count; ++i) {

    memcpy(ptr, &imgdata->blocks[i].blockmaxy, sizeof(uint8_t));
    ptr += 1;
    memcpy(ptr, &imgdata->blocks[i].blockminy, sizeof(uint8_t));
    ptr += 1;
    memcpy(ptr, &imgdata->blocks[i].blockmaxu, sizeof(uint8_t));
    ptr += 1;
    memcpy(ptr, &imgdata->blocks[i].blockminu, sizeof(uint8_t));
    ptr += 1;
    memcpy(ptr, &imgdata->blocks[i].blockmaxv, sizeof(uint8_t));
    ptr += 1;
    memcpy(ptr, &imgdata->blocks[i].blockminv, sizeof(uint8_t));
    ptr += 1;

    uint8_t interpolaten = (imgdata->blocks[i].interpolatey << 2) |
                           (imgdata->blocks[i].interpolateu << 1) |
                           imgdata->blocks[i].interpolatev;
    memcpy(ptr, &interpolaten, sizeof(uint8_t));
    ptr += sizeof(uint8_t);
  }

  for (int i = 0; i < imgdata->block_count; ++i) {
    memcpy(ptr, imgdata->blocks[i].corners, sizeof(uint8_t) * 4);
    ptr += sizeof(uint8_t) * 4;
  }

  for (int i = 0; i < imgdata->block_count; ++i) {
    memcpy(ptr, imgdata->blocks[i].nblock4bn, sizeof(uint8_t) * 64);
    ptr += sizeof(uint8_t) * 64;
  }

  memcpy(ptr, binfmt_endmsg, strlen(binfmt_endmsg));
  ptr += strlen(binfmt_endmsg);

  *buffer = buf;
  *size = total_size;
}

ImgData *buf_to_img(const char *buffer, size_t size) {
  const char *ptr = buffer;

  if (size < strlen(binfmt_msg) ||
      memcmp(ptr, binfmt_msg, strlen(binfmt_msg)) != 0) {
    fprintf(stderr, "Invalid header. %s\n", ptr);
    return NULL;
  }
  ptr += strlen(binfmt_msg);

  ImgData *img = (ImgData *)malloc(sizeof(ImgData));
  if (!img) {
    fprintf(stderr, "Memory allocation failed for ImgData.\n");
    return NULL;
  }

  int16_t width_be, height_be;
  memcpy(&width_be, ptr, sizeof(int16_t));
  ptr += sizeof(int16_t);
  memcpy(&height_be, ptr, sizeof(int16_t));
  ptr += sizeof(int16_t);
  img->width = ntohs(width_be);
  img->height = ntohs(height_be);

  int32_t block_count_be;
  memcpy(&block_count_be, ptr, sizeof(int32_t));
  ptr += sizeof(int32_t);
  img->block_count = ntohl(block_count_be);

  img->blocks = (BlockData *)malloc(sizeof(BlockData) * img->block_count);
  if (!img->blocks) {
    fprintf(stderr, "Memory allocation failed for BlockData.\n");
    free(img);
    return NULL;
  }

  for (int i = 0; i < img->block_count; ++i) {
    memcpy(&img->blocks[i].blockmaxy, ptr, sizeof(uint8_t) * 6);
    ptr += sizeof(uint8_t) * 6;
    uint8_t interpolaten;
    memcpy(&interpolaten, ptr, sizeof(uint8_t));
    ptr += sizeof(uint8_t);

    img->blocks[i].interpolatey = (interpolaten >> 2) & 1;
    img->blocks[i].interpolateu = (interpolaten >> 1) & 1;
    img->blocks[i].interpolatev = interpolaten & 1;
  }

  for (int i = 0; i < img->block_count; ++i) {
    memcpy(img->blocks[i].corners, ptr, sizeof(uint8_t) * 4);
    ptr += sizeof(uint8_t) * 4;
  }

  for (int i = 0; i < img->block_count; ++i) {
    memcpy(img->blocks[i].nblock4bn, ptr, sizeof(uint8_t) * 64);
    ptr += sizeof(uint8_t) * 64;
  }

  if (size < (ptr - buffer) + strlen(binfmt_endmsg) ||
      memcmp(ptr, binfmt_endmsg, strlen(binfmt_endmsg)) != 0) {
    fprintf(stderr, "Invalid footer.\n");
    free_imgdata(img);
    return NULL;
  }

  return img;
}

void free_imgdata(ImgData *imgdata) {
  if (imgdata) {
    if (imgdata->blocks) {
      free(imgdata->blocks);
    }
    free(imgdata);
  }
}